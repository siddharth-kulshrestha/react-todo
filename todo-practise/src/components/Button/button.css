.button-container {
    text-align: center;
    border-width: thin;
    border-style: solid;
    border-radius: 5%;
    border-color: black;
}