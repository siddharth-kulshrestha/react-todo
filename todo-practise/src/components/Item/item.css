.item-container {
    border-style: solid;
    border-radius: 5%;
    border-color: black;
    justify-content: space-around;
    align-items: center;
    flex-direction: row;
    display: flex;
    flex: 1;
    
}

.item-content {
    display: flex;
    flex-direction: row;
    flex: 2;
    margin: 10px;
}

.item-rate-handler {
    display: flex;
    flex-direction: row;
    flex: 1;
}

.rating-container {
    display: flex;
    flex-direction:row;
    justify-content: space-between;
}