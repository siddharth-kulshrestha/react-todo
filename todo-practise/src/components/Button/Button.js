import React from 'react';
import './button.css';

export const Button = ({ onClick, text }) => (
    <span className="button-container" onClick={onClick} >
        {text}
    </span>
);
