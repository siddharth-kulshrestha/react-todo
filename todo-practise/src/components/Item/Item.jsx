import React, { PureComponent } from 'react';
import './item.css';
import { Button } from '../Button/Button';

class RateItem extends PureComponent {

    state = {
        rating: 0,
    };

    componentDidMount() {
        if (this.props.rating !== undefined) {
            this.setState({rating: this.props.rating});
        }
    }

    onDecreaseRating() {
        this.setState((prevState) => ({rating: prevState.rating - 1}), this.onRatingChange);
    }

    onIncreaseRating() {
        this.setState((prevState) => ({rating: prevState.rating + 1}), this.onRatingChange);
    }

    onRatingChange() {
        if (this.props.onRatingChange) {
            this.props.onRatingChange(this.state.rating);
        }
    }

    render() {
        return (
            <div className="rating-container">
                <Button onClick={this.onDecreaseRating.bind(this)} text="-"/> {this.state.rating} <Button onClick={this.onIncreaseRating.bind(this)} text="+"/>
            </div>
        );
    }
}

export default class Item extends PureComponent {
    render() {
        console.log("Item-ID: "+this.props.data.id);
        console.log(this.props);
        return (
            <div className="item-container">
                <div className="item-content"><h3>{this.props.data.item.content}</h3></div> 
                <div className="item-rate-handler"><RateItem onRatingChange={(val) => {this.props.onRatingChange(this.props.data.item.id, val)}} rating={this.props.data.item.rating}/></div>
            </div>
        );
    }
}

