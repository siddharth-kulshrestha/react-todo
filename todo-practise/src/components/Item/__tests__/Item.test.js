import React from 'react';
import { shallow, configure, mount } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import Item, { RateItem } from '../Item';
import { Button } from '../../Button/Button';

configure({adapter: new Adapter()});

const mockItem = {
    id: 0,
    item: {
        content: "test content",
        rating: 6,
    }
};

// Snapshot test for Item component
describe('Item ', () => {
    it('should render correctly ', () => {
        const component = shallow(<Item data={mockItem} />);
        expect(component).toMatchSnapshot();
    });

    // Not really needed covered in snapshot testing
    it('should render content of item correctly ', () => {
        const component = shallow(<Item data={mockItem} />);
        const text = component.find('h3').text();
        expect(text).toEqual('test content');
    });
});

// Testing Item with RateItem component
describe('Testing RateItem component with Item component ', () => {

    it('Checking if rate component rate value is rendered correctly ', () => {
        const component = mount(<Item data={mockItem} />);
        const text = component.find(RateItem).find('b').text(); // Here we are accessing RateItem component so mount() is necessary
        expect(text).toEqual(' 6 ');
        component.unmount();
    });
    
    
    it(' Testing change of rate increase in item ', () => {
        const onRatingChangeFunc = jest.fn();
        const component = mount(<Item data={mockItem} onRatingChange={onRatingChangeFunc} />);
        component.find(RateItem).findWhere((comp) => comp.type() == Button && comp.text() == '+').simulate('click');
        expect(onRatingChangeFunc).toBeCalledWith(0, 7);
        component.unmount();
    });

    it(' Testing change of rate decrease in item ', () => {
        const onRatingChangeFunc = jest.fn();
        const component = mount(<Item data={mockItem} onRatingChange={onRatingChangeFunc} />);
        component.find(RateItem).findWhere((comp) => comp.type() == Button && comp.text() == '-').simulate('click');
        expect(onRatingChangeFunc).toBeCalledWith(0, 5);
        component.unmount();
    });
});
