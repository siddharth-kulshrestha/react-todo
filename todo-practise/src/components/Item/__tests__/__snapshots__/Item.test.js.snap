// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Item  should render correctly  1`] = `
<div
  className="item-container"
>
  <div
    className="item-content"
  >
    <h3>
      test content
    </h3>
  </div>
  <div
    className="item-rate-handler"
  >
    <RateItem
      onRatingChange={[Function]}
      rating={6}
    />
  </div>
</div>
`;
